docs/source/dev-guide/techspec-solver-state.md

```diff
@@ -0,0 +1,159 @@
+a given set of initial conditions. At least we'll try. Take into account that the `specs` list
+is kept around across attempts! In other words, the `specs` list is only really empty in the first
+attempt; if this fails, the subsequent attempts will only overwrite (update) the existing one. In
+practice, this should only affect how constrained packages are. The names should be the same.
+
+It will also depend on whether we are adding (`conda install|create|update`) or removing (
+`conda remove`) packages. There's a common initialization part for both, but after that the
+logic is separate.
+
+<!--
+I will first describe what we are doing step by step. Hopefully by writing this down we can
+think of potential simplifications of the logic and its implementation.
+-->
+
+# Common initialization
+
+> Note: This happens in `Solver._collect_all_metadata()`
+
+This happens regardless the type of command we are using (`install`, `update`, `create` or
+`remove`).
+
+1. Add specs from `history`, if any.
+2. Add specs from `do_not_remove`, but only if:
+    * There's no spec for that name in `specs` already, _and_
+    * A package with that name is _not_ installed.
+3. Add `virtual` packages as unconstrained `specs`.
+4. Add all those installed packages, as unconstrained `specs`, that satisfy any of these conditions:
+    * The history is empty (in that case all installed packages are added)
+    * The package name is part of `aggresive_updates`
+    * The package was no installed by `conda`, but by pip or other PyPI tools.
+
+<!--
+This is getting simplified in the libmamba refactor, as the following. It should be equivalent:
+
+If we have a history:
+    * Add history specs
+    * Add installed packages as unconstrained specs that satisfy any of these conditions:
+        * Part of aggresive updates
+        * Part of do_not_remove
+        * Installed by pip
+Else, we add _all_ installed packages as unconstrained specs.
+
+Finally, add virtual packages as bare specs.
+-->
+
+```{admonition} Preparing the index
+At this point, the populated `specs` and the `requested` specs are merged together. This temporary
+collection is used to determine how to reduce the index.
+```
+# Processing specs for `conda install`
+
+## Preparation
+
+1. Generate the explicit pool for the requested specs (via `Resolve._get_package_pool()`).
+2. Detect potential conflicts (via (`Resolve.get_conflicting_specs()`).
+
+## Refine `specs` that match installed records
+
+1. Check that each of `specs` match a single installed package or none! Otherwise (two or more
+   matches), it means that environment is in bad shape and basically broken. If the `spec` matches
+   one installed package, let's call it _installed match_, we will modify the original spec.
+2. We will turn the `spec` into an _exact_ (frozen) spec if:
+    1. If the installed match is unmanageable (installed by pip, virtual, etc)
```

Comment:
You don't need the "If" at the beginning of item 1 if you're putting it at the end of the line above.

Also, "etc" needs a period after it.